src/com/google/javascript/jscomp/ProcessCommonJSModules.java

```diff
@@ -943,66 +1138,95 @@ private void visitExport(NodeTraversal t, Node export) {
             && rValue.isObjectLit()
             && root.getParent().isAssign()
             && root.getParent().getParent().isExprResult()) {
-          expandObjectLitAssignment(t, root);
+          expandObjectLitAssignment(t, root, export.scope);
           return;
         }
       }
 
-      ModulePath modulePath = t.getInput().getPath();
-      String moduleName = modulePath.toModuleName();
+      String moduleName = getModuleName(t.getInput());
+      Var moduleInitialization = t.getScope().getVar(moduleName);
 
       // If this is an assignment to module.exports or exports, renaming
       // has already handled this case. Remove the export.
       Var rValueVar = null;
       if (rValue != null && rValue.isQualifiedName()) {
-        rValueVar = t.getScope().getVar(rValue.getQualifiedName());
-
-        // If the exported name is not found and this is a direct assignment
-        // to modules.exports, look to see if the module name has a var definition
-        if (rValueVar == null && root == export) {
-          rValueVar = t.getScope().getVar(moduleName);
-          if (rValueVar != null && rValueVar.getNode() == root) {
-            rValueVar = null;
-          }
-        }
+        rValueVar = export.scope.getVar(rValue.getQualifiedName());
       }
 
       if (root.getParent().isAssign()
           && (root.getNext() != null && (root.getNext().isName() || root.getNext().isGetProp()))
           && root.getParent().getParent().isExprResult()
-          && rValueVar != null) {
+          && rValueVar != null
+          && (NodeUtil.getEnclosingScript(rValueVar.nameNode) == null
+              || (rValueVar.nameNode.getParent() != null && !rValueVar.isParam()))) {
         root.getParent().getParent().detach();
         t.reportCodeChange();
         return;
       }
 
+      moduleName = moduleName + "." + EXPORT_PROPERTY_NAME;
+
       Node updatedExport =
-          NodeUtil.newName(compiler, moduleName, export, export.getQualifiedName());
+          NodeUtil.newQName(compiler, moduleName, export.node, export.node.getQualifiedName());
+      boolean exportIsConst =
+          defaultExportIsConst
+              && getBasePropertyImport(getModuleName(t.getInput()))
```

Comment:
Can be replaced with simpler, equivalent code.
*Caveats*: This version no longer throws NPE if the first object in comparison is null.

Suggested replacement:
```java
              && updatedExport.matchesQualifiedName(
                  getBasePropertyImport(getModuleName(t.getInput())))
```